Feature/asm syntax

berkus opened the pull request:
Switch from llvm_asm!() syntax to asm!() syntax.

Add register width specification for future expansion.

Since asm!() is not fully stabilised this might need to wait, but I still want to push it out from my PR closet.

berkus committed:
Switch barrier operations from llvm_asm! to asm!

andre-richter commented:
Hi @berkus,

I am unsure what the gain / idea behind the explicit register width modifiers is.
We are reading specific registers everywhere, no placeholders are used. So the register width modifier is redundant, or am I wrong?

Also, this PR would clash with the current state of #21 afaics.
Ideally, I would want both of the PRs merged. My proposal would be to clean up this one by removing the width specifier for now, unless I misunderstood it, and then we can get both in and do a release?

berkus commented:
Register width specifier is for better extensibility, for example aarch32 insns might use "w" instead of "x".

In my copy I have both asm!() syntax as well as #21 applied, so I can rebase either of them to use preferred syntax easily. I'd probably merge 21 first. 

What I'm worried about is if we hardcode the register specifiers as `asm!(concat!($asm_instr, " {reg:x}, ", $asm_reg_name), reg = out(reg) reg, options(nomem, nostack));` it will make it harder to reuse these macros (see #18 too).

If we don't specify the widths, compiler has hard time guessing what the correct width should be:
```
warning: formatting may not be suitable for sub-register argument
  --> src/regs/macros.rs:18:30
   |
18 |                         asm!(concat!($asm_instr, " {reg}, ", $asm_reg_name), reg = out(reg) reg, options(nomem, nostack));
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                 --- for this argument
   |
  ::: src/regs/cntfrq_el0.rs:19:5
   |
19 |     sys_coproc_read_raw!(u32, "CNTFRQ_EL0", "x");
   |     --------------------------------------------- in this macro invocation
   |
   = note: `#[warn(asm_sub_register)]` on by default
   = help: use the `w` modifier to have the register formatted as `w0`
   = help: or use the `x` modifier to keep the default formatting of `x0`
   = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Explanation says:
```
Registers on some architectures can use different names to refer to a subset of the register. By default, the compiler will use the name for the full register size.
```

So it will use "x" by default, but this may not be suitable for all occasions in the future.

berkus commented:
If it's ok, merge #21 first and then I'll update #20 to work with that?

andre-richter commented:
Thanks for the explanation. I wasn't aware those warnings would be emitted. Haven't touched much of the new syntax yet.

#21 first it is, then 👍

berkus committed:
Switch register macros from llvm_asm! to asm!

Add register width specifier for future expansion.

berkus committed:
Add MSRV notice

berkus commented:
I've rebased and added a rustc version notice (unfortunately, the rustc release notes contain no information about when asm!() was introduced, but I tracked it to 1.45.0).
